From f4e40e6b8f031e071a8506738a0208aa4e4c8f54 Mon Sep 17 00:00:00 2001
From: Jason Zeng <jason.zeng@intel.com>
Date: Fri, 7 Oct 2022 13:47:19 +0800
Subject: [PATCH 0288/1646] [Backport] x86/cpu: fix kabi for
 cpuinfo_x86.vmx_capability

Intel inclusion
category: feature
bugzilla: https://gitee.com/openeuler/intel-kernel/issues/I5ODSC

Reference: https://gitee.com/openeuler/kernel/commit/f0af26c88e3f949f174d5a180b46823ed37c3be6

-------------------------------------------------

The introduction of VMX tertiary features like IPI virtualization
causes the change of the size of struct cpu_info_x86. This patch
tries to put the tertiary features on a separate data structure.

Signed-off-by: Jason Zeng <jason.zeng@intel.com>
Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>
---
 arch/x86/include/asm/processor.h   | 16 ++++++++++++++++
 arch/x86/include/asm/vmx.h         |  3 ++-
 arch/x86/include/asm/vmxfeatures.h |  5 +++--
 arch/x86/kernel/cpu/common.c       | 10 ++++++++++
 arch/x86/kernel/cpu/feat_ctl.c     | 25 ++++++++++++++++++-------
 arch/x86/kernel/cpu/mkcapflags.sh  |  3 +++
 arch/x86/kernel/cpu/proc.c         |  8 ++++++++
 arch/x86/kernel/setup.c            |  2 ++
 arch/x86/kernel/smpboot.c          |  2 ++
 9 files changed, 64 insertions(+), 10 deletions(-)

diff --git a/arch/x86/include/asm/processor.h b/arch/x86/include/asm/processor.h
index c55ab98f2d2f..d8335c680998 100644
--- a/arch/x86/include/asm/processor.h
+++ b/arch/x86/include/asm/processor.h
@@ -142,6 +142,12 @@ struct cpuinfo_x86 {
 	unsigned		initialized : 1;
 } __randomize_layout;
 
+struct extra_cpuinfo_x86 {
+#ifdef CONFIG_X86_VMX_FEATURE_NAMES
+	__u32			vmx_tertiary_capability[NVMX_TERTIARY_INTS];
+#endif
+} __randomize_layout;
+
 struct cpuid_regs {
 	u32 eax, ebx, ecx, edx;
 };
@@ -172,6 +178,8 @@ enum cpuid_regs_idx {
 extern struct cpuinfo_x86	boot_cpu_data;
 extern struct cpuinfo_x86	new_cpu_data;
 
+extern struct extra_cpuinfo_x86	extra_boot_cpu_data;
+
 extern __u32			cpu_caps_cleared[NCAPINTS + NBUGINTS];
 extern __u32			cpu_caps_set[NCAPINTS + NBUGINTS];
 
@@ -183,6 +191,14 @@ DECLARE_PER_CPU_READ_MOSTLY(struct cpuinfo_x86, cpu_info);
 #define cpu_data(cpu)		boot_cpu_data
 #endif
 
+#ifdef CONFIG_SMP
+DECLARE_PER_CPU_READ_MOSTLY(struct extra_cpuinfo_x86, extra_cpu_info);
+#define extra_cpu_data(cpu)	per_cpu(extra_cpu_info, cpu)
+#else
+#define extra_cpu_info		extra_boot_cpu_data
+#define extra_cpu_data(cpu)	extra_boot_cpu_data
+#endif
+
 extern const struct seq_operations cpuinfo_op;
 
 #define cache_line_size()	(boot_cpu_data.x86_cache_alignment)
diff --git a/arch/x86/include/asm/vmx.h b/arch/x86/include/asm/vmx.h
index aaf3812063bb..ecae8a49e600 100644
--- a/arch/x86/include/asm/vmx.h
+++ b/arch/x86/include/asm/vmx.h
@@ -18,6 +18,7 @@
 #include <asm/vmxfeatures.h>
 
 #define VMCS_CONTROL_BIT(x)	BIT(VMX_FEATURE_##x & 0x1f)
+#define VMCS_TERTIARY_CONTROL_BIT(x)	BIT(VMX_TERTIARY_FEATURE_##x & 0x1f)
 
 /*
  * Definitions of Primary Processor-Based VM-Execution Controls.
@@ -78,7 +79,7 @@
 /*
  * Definitions of Tertiary Processor-Based VM-Execution Controls.
  */
-#define TERTIARY_EXEC_IPI_VIRT			VMCS_CONTROL_BIT(IPI_VIRT)
+#define TERTIARY_EXEC_IPI_VIRT			VMCS_TERTIARY_CONTROL_BIT(IPI_VIRT)
 
 #define PIN_BASED_EXT_INTR_MASK                 VMCS_CONTROL_BIT(INTR_EXITING)
 #define PIN_BASED_NMI_EXITING                   VMCS_CONTROL_BIT(NMI_EXITING)
diff --git a/arch/x86/include/asm/vmxfeatures.h b/arch/x86/include/asm/vmxfeatures.h
index af446ba041db..f26dea5769d8 100644
--- a/arch/x86/include/asm/vmxfeatures.h
+++ b/arch/x86/include/asm/vmxfeatures.h
@@ -5,7 +5,8 @@
 /*
  * Defines VMX CPU feature bits
  */
-#define NVMXINTS			5 /* N 32-bit words worth of info */
+#define NVMXINTS			3 /* N 32-bit words worth of info */
+#define NVMX_TERTIARY_INTS		2 /* N 32-bit words worth of info */
 
 /*
  * Note: If the comment begins with a quoted string, that string is used
@@ -86,5 +87,5 @@
 #define VMX_FEATURE_ENCLV_EXITING	( 2*32+ 28) /* "" VM-Exit on ENCLV (leaf dependent) */
 
 /* Tertiary Processor-Based VM-Execution Controls, word 3 */
-#define VMX_FEATURE_IPI_VIRT		( 3*32+  4) /* Enable IPI virtualization */
+#define VMX_TERTIARY_FEATURE_IPI_VIRT		( 3*32+  4) /* Enable IPI virtualization */
 #endif /* _ASM_X86_VMXFEATURES_H */
diff --git a/arch/x86/kernel/cpu/common.c b/arch/x86/kernel/cpu/common.c
index 70bbabbde25b..17414fa6e71e 100644
--- a/arch/x86/kernel/cpu/common.c
+++ b/arch/x86/kernel/cpu/common.c
@@ -1613,6 +1613,15 @@ static void validate_apic_and_package_id(struct cpuinfo_x86 *c)
 #endif
 }
 
+#ifdef CONFIG_X86_VMX_FEATURE_NAMES
+static void init_extra_cpu_data(u16 cpu_index)
+{
+	struct extra_cpuinfo_x86 *e = &extra_cpu_data(cpu_index);
+
+	memset(&e->vmx_tertiary_capability, 0, sizeof(e->vmx_tertiary_capability));
+}
+#endif
+
 /*
  * This does the hard work of actually picking apart the CPU stuff...
  */
@@ -1643,6 +1652,7 @@ static void identify_cpu(struct cpuinfo_x86 *c)
 	memset(&c->x86_capability, 0, sizeof(c->x86_capability));
 #ifdef CONFIG_X86_VMX_FEATURE_NAMES
 	memset(&c->vmx_capability, 0, sizeof(c->vmx_capability));
+	init_extra_cpu_data(c->cpu_index);
 #endif
 
 	generic_identify(c);
diff --git a/arch/x86/kernel/cpu/feat_ctl.c b/arch/x86/kernel/cpu/feat_ctl.c
index 03851240c3e3..bd86b7b55265 100644
--- a/arch/x86/kernel/cpu/feat_ctl.c
+++ b/arch/x86/kernel/cpu/feat_ctl.c
@@ -15,16 +15,30 @@ enum vmx_feature_leafs {
 	MISC_FEATURES = 0,
 	PRIMARY_CTLS,
 	SECONDARY_CTLS,
-	TERTIARY_CTLS_LOW,
-	TERTIARY_CTLS_HIGH,
 	NR_VMX_FEATURE_WORDS,
 };
 
+enum vmx_tertiary_feature_leafs {
+	TERTIARY_CTLS_LOW = 0,
+	TERTIARY_CTLS_HIGH,
+};
+
 #define VMX_F(x) BIT(VMX_FEATURE_##x & 0x1f)
 
+static void init_vmx_tertiary_capabilities(u16 cpu_index)
+{
+	struct extra_cpuinfo_x86 *e = &extra_cpu_data(cpu_index);
+	u32 low, high;
+
+	/* All 64 bits of tertiary controls MSR are allowed-1 settings. */
+	rdmsr_safe(MSR_IA32_VMX_PROCBASED_CTLS3, &low, &high);
+	e->vmx_tertiary_capability[TERTIARY_CTLS_LOW] = low;
+	e->vmx_tertiary_capability[TERTIARY_CTLS_HIGH] = high;
+}
+
 static void init_vmx_capabilities(struct cpuinfo_x86 *c)
 {
-	u32 supported, funcs, ept, vpid, ign, low, high;
+	u32 supported, funcs, ept, vpid, ign;
 
 	BUILD_BUG_ON(NVMXINTS != NR_VMX_FEATURE_WORDS);
 
@@ -44,10 +58,7 @@ static void init_vmx_capabilities(struct cpuinfo_x86 *c)
 	rdmsr_safe(MSR_IA32_VMX_PROCBASED_CTLS2, &ign, &supported);
 	c->vmx_capability[SECONDARY_CTLS] = supported;
 
-	/* All 64 bits of tertiary controls MSR are allowed-1 settings. */
-	rdmsr_safe(MSR_IA32_VMX_PROCBASED_CTLS3, &low, &high);
-	c->vmx_capability[TERTIARY_CTLS_LOW] = low;
-	c->vmx_capability[TERTIARY_CTLS_HIGH] = high;
+	init_vmx_tertiary_capabilities(c->cpu_index);
 
 	rdmsr(MSR_IA32_VMX_PINBASED_CTLS, ign, supported);
 	rdmsr_safe(MSR_IA32_VMX_VMFUNC, &ign, &funcs);
diff --git a/arch/x86/kernel/cpu/mkcapflags.sh b/arch/x86/kernel/cpu/mkcapflags.sh
index 1db560ed2ca3..b030882d2f6f 100644
--- a/arch/x86/kernel/cpu/mkcapflags.sh
+++ b/arch/x86/kernel/cpu/mkcapflags.sh
@@ -68,6 +68,9 @@ trap 'rm "$OUT"' EXIT
 	echo "#include <asm/vmxfeatures.h>"
 	echo "#endif"
 	dump_array "x86_vmx_flags" "NVMXINTS*32" "VMX_FEATURE_" "" $3
+	echo ""
+
+	dump_array "x86_vmx_tertiary_flags" "NVMX_TERTIARY_INTS*32" "VMX_TERTIARY_FEATURE_" "NVMXINTS*32" $3
 	echo "#endif /* CONFIG_X86_VMX_FEATURE_NAMES */"
 ) > $OUT
 
diff --git a/arch/x86/kernel/cpu/proc.c b/arch/x86/kernel/cpu/proc.c
index 4eec8889b0ff..7810c75abdd4 100644
--- a/arch/x86/kernel/cpu/proc.c
+++ b/arch/x86/kernel/cpu/proc.c
@@ -1,4 +1,5 @@
 // SPDX-License-Identifier: GPL-2.0
+#include "asm/vmxfeatures.h"
 #include <linux/smp.h>
 #include <linux/timex.h>
 #include <linux/string.h>
@@ -9,6 +10,7 @@
 
 #ifdef CONFIG_X86_VMX_FEATURE_NAMES
 extern const char * const x86_vmx_flags[NVMXINTS*32];
+extern const char * const x86_vmx_tertiary_flags[NVMX_TERTIARY_INTS*32];
 #endif
 
 /*
@@ -108,12 +110,18 @@ static int show_cpuinfo(struct seq_file *m, void *v)
 
 #ifdef CONFIG_X86_VMX_FEATURE_NAMES
 	if (cpu_has(c, X86_FEATURE_VMX) && c->vmx_capability[0]) {
+		struct extra_cpuinfo_x86 *e = &extra_cpu_data(c->cpu_index);
 		seq_puts(m, "\nvmx flags\t:");
 		for (i = 0; i < 32*NVMXINTS; i++) {
 			if (test_bit(i, (unsigned long *)c->vmx_capability) &&
 			    x86_vmx_flags[i] != NULL)
 				seq_printf(m, " %s", x86_vmx_flags[i]);
 		}
+		for (i = 0; i < 32*NVMX_TERTIARY_INTS; i++) {
+			if (test_bit(i, (unsigned long *)e->vmx_tertiary_capability) &&
+			    x86_vmx_tertiary_flags[i] != NULL)
+				seq_printf(m, " %s", x86_vmx_tertiary_flags[i]);
+		}
 	}
 #endif
 
diff --git a/arch/x86/kernel/setup.c b/arch/x86/kernel/setup.c
index 85979c1a404e..4d48ebb18fe8 100644
--- a/arch/x86/kernel/setup.c
+++ b/arch/x86/kernel/setup.c
@@ -5,6 +5,7 @@
  * This file contains the setup_arch() code, which handles the architecture-dependent
  * parts of early kernel initialization.
  */
+#include "asm/processor.h"
 #include <linux/console.h>
 #include <linux/crash_dump.h>
 #include <linux/dma-map-ops.h>
@@ -141,6 +142,7 @@ struct cpuinfo_x86 boot_cpu_data __read_mostly;
 EXPORT_SYMBOL(boot_cpu_data);
 #endif
 
+struct extra_cpuinfo_x86 extra_boot_cpu_data __read_mostly;
 
 #if !defined(CONFIG_X86_PAE) || defined(CONFIG_X86_64)
 __visible unsigned long mmu_cr4_features __ro_after_init;
diff --git a/arch/x86/kernel/smpboot.c b/arch/x86/kernel/smpboot.c
index bda89ecc7799..57c3f0a25c0a 100644
--- a/arch/x86/kernel/smpboot.c
+++ b/arch/x86/kernel/smpboot.c
@@ -99,6 +99,8 @@ DEFINE_PER_CPU_READ_MOSTLY(cpumask_var_t, cpu_llc_shared_map);
 /* Per CPU bogomips and other parameters */
 DEFINE_PER_CPU_READ_MOSTLY(struct cpuinfo_x86, cpu_info);
 EXPORT_PER_CPU_SYMBOL(cpu_info);
+DEFINE_PER_CPU_READ_MOSTLY(struct extra_cpuinfo_x86, extra_cpu_info);
+EXPORT_PER_CPU_SYMBOL(extra_cpu_info);
 
 /* Logical package management. We might want to allocate that dynamically */
 unsigned int __max_logical_packages __read_mostly;
-- 
2.17.1

